lib/timeutils/conv.c

```diff
@@ -0,0 +1,112 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include "timeutils/conv.h"
+#include "timeutils/cache.h"
+#include "timeutils/misc.h"
+
+void
+convert_wall_clock_time_to_unix_time(UnixTime *self, const WallClockTime *wct)
+{
+  WallClockTime work_wct = *wct;
+  convert_and_normalize_wall_clock_time_to_unix_time(self, &work_wct);
+}
+
+void
+convert_wall_clock_time_to_unix_time_with_tz_hint(UnixTime *self, const WallClockTime *wct, long gmtoff_hint)
+{
+  WallClockTime work_wct = *wct;
+  convert_and_normalize_wall_clock_time_to_unix_time_with_tz_hint(self, &work_wct, gmtoff_hint);
+}
+
+void
+convert_and_normalize_wall_clock_time_to_unix_time(UnixTime *self, WallClockTime *wct)
+{
+  convert_and_normalize_wall_clock_time_to_unix_time_with_tz_hint(self, wct, -1);
+}
+
+/* hint the timezone value if it is not present in the wct struct, e.g.  the
+ * timestamp takes precedence, but as an additional information the caller
+ * can supply its best idea.  this maps nicely to the source-side
+ * time-zone() value, which is only used in case an incoming timestamp does
+ * not have the timestamp value. */
+void
+convert_and_normalize_wall_clock_time_to_unix_time_with_tz_hint(UnixTime *self, WallClockTime *wct, long gmtoff_hint)
+{
+  gint target_gmtoff = -1;
+
+  /* usec is just copied over, doesn't change timezone or anything */
+  self->ut_usec = wct->wct_usec;
+
+  /* determine target gmtoff if it's coming from the timestamp or from the hint */
+  target_gmtoff = wct->wct_gmtoff;
+  if (target_gmtoff == -1)
+    target_gmtoff = gmtoff_hint;
+
+  /* FIRST: We convert the timestamp as it was in our local time zone. */
+  gint unnormalized_hour = wct->wct_hour;
+  wct->wct_isdst = -1;
+  self->ut_sec = cached_mktime_wct(wct);
+  gint normalized_hour = wct->wct_hour;
+
+
+  /* SECOND: adjust ut_sec as if we converted it according to our timezone. */
+  gint local_gmtoff = get_local_timezone_ofs(self->ut_sec);
+  if (target_gmtoff == -1)
+    {
+      target_gmtoff = local_gmtoff;
+    }
+  self->ut_sec = self->ut_sec
+                 + local_gmtoff
+                 - (normalized_hour - unnormalized_hour) * 3600
```

Comment:
Can you explain why we need to repair with `- (normalized_hour - unnormalized_hour) * 3600`? `mktime` should deal with aligning calendar time due to normalization before returning. I think we do not need to repair here.

Reply:
We do, as we are using local timezone specific functions (e.g. mktime) to convert timestamps in a remote timezone.

What we do is that we convert these timestamps and then cancel out any local timezone offsets. And the normalized/unnormalized dance is needed for daylight saving transition hours. The US changes dst at a different time than in Europe, so if mktime() uses that to change the local hour value, then that is incorrect.

Also, this branch explicitly does not want to change behavior, and this is the same as the one before.

Reply:
Oh I see. I was thinking this is a new implementation, then I found `_adjust_tvsec_to_move_it_into_given_timezone` which worked exactly this way. I think I understand this one.